// express/confluence-writer.ts
import { createChildLogger } from '../utils/logger.js';
import { parseConfluenceUrl, ConfluenceError } from './confluence-reader.js';

const log = createChildLogger({ module: 'confluence-writer' });

// ─── Types ───

export interface ScreenshotData {
  name: string;
  base64: string;
  figmaLink: string;
}

export interface UpdatePrdResult {
  success: boolean;
  updatedCount: number;
  pageUrl: string;
  newPageId?: string;
  error?: string;
}

export interface ExistingImageInfo {
  index: number;
  imageHtml: string;
  imagePreviewUrl: string | null;
  figmaLink: string | null;
}

export interface PreviewImagesResult {
  success: boolean;
  images: ExistingImageInfo[];
  pageTitle: string;
  error?: string;
}

interface ConfluencePageInfo {
  id: string;
  title: string;
  version: number;
  spaceKey: string;
  body: string;
}

// ─── Confluence API ───

/**
 * 获取 Confluence 页面信息（包括版本号和内容）
 */
async function getPageInfo(
  baseUrl: string,
  pageId: string,
  pat: string,
): Promise<ConfluencePageInfo> {
  const apiUrl = `${baseUrl}/rest/api/content/${pageId}?expand=body.storage,version,space`;

  const response = await fetch(apiUrl, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${pat}`,
      'Accept': 'application/json',
    },
    signal: AbortSignal.timeout(15_000),
  });

  if (!response.ok) {
    const status = response.status;
    if (status === 401) throw new ConfluenceError('AUTH', 'PAT 无效或已过期');
    if (status === 403) throw new ConfluenceError('FORBIDDEN', '无权限访问该页面');
    if (status === 404) throw new ConfluenceError('NOT_FOUND', '页面不存在');
    throw new ConfluenceError('API_ERROR', `Confluence API 错误 (${status})`);
  }

  const data = await response.json() as {
    id: string;
    title: string;
    version: { number: number };
    space: { key: string };
    body?: { storage?: { value?: string } };
  };

  return {
    id: data.id,
    title: data.title,
    version: data.version.number,
    spaceKey: data.space.key,
    body: data.body?.storage?.value || '',
  };
}

/**
 * 上传附件到 Confluence 页面
 */
async function uploadAttachment(
  baseUrl: string,
  pageId: string,
  pat: string,
  filename: string,
  base64Data: string,
): Promise<string> {
  const apiUrl = `${baseUrl}/rest/api/content/${pageId}/child/attachment`;

  const binaryData = Buffer.from(base64Data, 'base64');

  const boundary = '----WebKitFormBoundary' + Math.random().toString(36).substring(2);
  const bodyParts: string[] = [];

  bodyParts.push(`--${boundary}`);
  bodyParts.push(`Content-Disposition: form-data; name="file"; filename="${filename}"`);
  bodyParts.push('Content-Type: image/png');
  bodyParts.push('');

  const headerBuffer = Buffer.from(bodyParts.join('\r\n') + '\r\n', 'utf8');
  const footerBuffer = Buffer.from(`\r\n--${boundary}--\r\n`, 'utf8');
  const fullBody = Buffer.concat([headerBuffer, binaryData, footerBuffer]);

  const response = await fetch(apiUrl, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${pat}`,
      'X-Atlassian-Token': 'nocheck',
      'Content-Type': `multipart/form-data; boundary=${boundary}`,
    },
    body: fullBody,
    signal: AbortSignal.timeout(30_000),
  });

  if (!response.ok) {
    const errText = await response.text().catch(() => '');
    log.warn({ status: response.status, pageId, errText: errText.slice(0, 200) }, 'Failed to upload attachment');
    throw new ConfluenceError('API_ERROR', `上传附件失败 (${response.status})`);
  }

  log.info({ pageId, filename }, 'Attachment uploaded');
  return filename;
}

/**
 * 更新 Confluence 页面内容
 */
async function updatePageContent(
  baseUrl: string,
  pageId: string,
  pat: string,
  title: string,
  newBody: string,
  currentVersion: number,
): Promise<void> {
  const apiUrl = `${baseUrl}/rest/api/content/${pageId}`;

  const response = await fetch(apiUrl, {
    method: 'PUT',
    headers: {
      'Authorization': `Bearer ${pat}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      version: { number: currentVersion + 1 },
      type: 'page',
      title: title,
      body: {
        storage: {
          value: newBody,
          representation: 'storage',
        },
      },
    }),
    signal: AbortSignal.timeout(15_000),
  });

  if (!response.ok) {
    const errText = await response.text().catch(() => '');
    log.warn({ status: response.status, pageId, errText: errText.slice(0, 200) }, 'Failed to update page');
    throw new ConfluenceError('API_ERROR', `更新页面失败 (${response.status})`);
  }

  log.info({ pageId, title }, 'Page content updated');
}

/**
 * 创建子页面（复制原页面内容）
 */
async function createChildPage(
  baseUrl: string,
  parentPageId: string,
  spaceKey: string,
  pat: string,
  title: string,
  body: string,
): Promise<{ id: string; url: string }> {
  const apiUrl = `${baseUrl}/rest/api/content`;

  const response = await fetch(apiUrl, {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${pat}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      type: 'page',
      title: title,
      space: { key: spaceKey },
      ancestors: [{ id: parentPageId }],
      body: {
        storage: {
          value: body,
          representation: 'storage',
        },
      },
    }),
    signal: AbortSignal.timeout(15_000),
  });

  if (!response.ok) {
    const errText = await response.text().catch(() => '');
    log.warn({ status: response.status, parentPageId, errText: errText.slice(0, 200) }, 'Failed to create child page');
    throw new ConfluenceError('API_ERROR', `创建子页面失败 (${response.status})`);
  }

  const data = await response.json() as {
    id: string;
    _links: { webui: string };
  };

  const pageUrl = `${baseUrl}${data._links.webui}`;
  log.info({ parentPageId, newPageId: data.id, title }, 'Child page created');

  return { id: data.id, url: pageUrl };
}

// ─── Image Replacement Logic ───

/**
 * 从 Figma 链接 HTML 中提取 URL（支持 <a href> 和 <ac:link><ri:url> 格式）
 */
function extractFigmaUrl(linkHtml: string): string | null {
  // <a href="...figma...">
  const hrefMatch = /href="([^"]*figma[^"]*)"/i.exec(linkHtml);
  if (hrefMatch) return hrefMatch[1].replace(/&amp;/g, '&');

  // <ri:url ri:value="...figma..."/>
  const riValueMatch = /ri:value="([^"]*figma[^"]*)"/i.exec(linkHtml);
  if (riValueMatch) return riValueMatch[1].replace(/&amp;/g, '&');

  return null;
}

function extractAttachmentFilename(imageHtml: string): string | null {
  const match = /ri:filename="([^"]+)"/i.exec(imageHtml);
  return match ? match[1] : null;
}

function extractImageSrc(imageHtml: string): string | null {
  const srcMatch = /src="([^"]+)"/i.exec(imageHtml);
  if (srcMatch && srcMatch[1]) return srcMatch[1];

  const valueMatch = /ri:value="([^"]+)"/i.exec(imageHtml);
  return valueMatch ? valueMatch[1] : null;
}

function buildAttachmentPreviewUrl(baseUrl: string, pageId: string, filename: string): string {
  return `${baseUrl}/download/attachments/${pageId}/${encodeURIComponent(filename)}?api=v2`;
}

function escapeXmlText(value: string): string {
  return String(value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

function escapeXmlAttr(value: string): string {
  return escapeXmlText(value).replace(/"/g, '&quot;');
}

function findParagraphStart(html: string, index: number): number {
  const start = html.lastIndexOf('<p', index);
  return start;
}

function findParagraphEnd(html: string, index: number): number {
  const end = html.indexOf('</p>', index);
  return end === -1 ? -1 : end + 4;
}

function findContainingParagraphRange(html: string, index: number, matchLength: number): { start: number; end: number } | null {
  const start = findParagraphStart(html, index);
  const end = findParagraphEnd(html, index + matchLength);
  if (start === -1 || end === -1) return null;
  if (start > index || end < index + matchLength) return null;
  return { start, end };
}

/**
 * 查找图片后紧跟的 Figma 链接块。
 * Confluence 实际内容里，图片后经常会先出现 `</span></p>` 之类的收尾标签，
 * 再跟一个包含 figma 链接的 `<p>...</p>`，这里一并吃掉。
 */
function findFollowingFigmaLink(html: string, startIndex: number): string | null {
  const remaining = html.slice(startIndex);
  const prefixMatch = /^(?:\s|<\/[^>]+>)+/i.exec(remaining);
  const prefix = prefixMatch ? prefixMatch[0] : '';
  const trimmed = remaining.slice(prefix.length);

  const paragraphMatch = /^<p[\s\S]*?<\/p>/i.exec(trimmed);
  if (paragraphMatch && /figma/i.test(paragraphMatch[0])) {
    return prefix + paragraphMatch[0];
  }

  const acLinkMatch = /^<ac:link[\s\S]*?<\/ac:link>/i.exec(trimmed);
  if (acLinkMatch && /figma/i.test(acLinkMatch[0])) {
    return prefix + acLinkMatch[0];
  }

  const anchorMatch = /^<a\b[\s\S]*?<\/a>/i.exec(trimmed);
  if (anchorMatch && /figma/i.test(anchorMatch[0])) {
    return prefix + anchorMatch[0];
  }

  return null;
}

/**
 * 解析 HTML 中的图片位置（包含紧跟的 Figma 链接）
 * 支持 Confluence Storage Format 的 ac:image 和标准 img 标签
 */
function findImagePositions(html: string): Array<{ match: string; index: number }> {
  const images: Array<{ match: string; index: number }> = [];

  // 匹配 ac:image 宏
  const acImageRegex = /<ac:image[^>]*>[\s\S]*?<\/ac:image>/gi;
  let match: RegExpExecArray | null;
  while ((match = acImageRegex.exec(html)) !== null) {
    const imageRange = findContainingParagraphRange(html, match.index, match[0].length);
    const baseStart = imageRange ? imageRange.start : match.index;
    const baseEnd = imageRange ? imageRange.end : match.index + match[0].length;

    let fullMatch = html.slice(baseStart, baseEnd);
    const figmaLink = findFollowingFigmaLink(html, baseEnd);
    if (figmaLink) fullMatch += figmaLink;

    images.push({ match: fullMatch, index: baseStart });
  }

  // 匹配 img 标签
  const imgRegex = /<img[^>]*\/?>/gi;
  while ((match = imgRegex.exec(html)) !== null) {
    const imageRange = findContainingParagraphRange(html, match.index, match[0].length);
    const baseStart = imageRange ? imageRange.start : match.index;
    const baseEnd = imageRange ? imageRange.end : match.index + match[0].length;

    let fullMatch = html.slice(baseStart, baseEnd);
    const figmaLink = findFollowingFigmaLink(html, baseEnd);
    if (figmaLink) fullMatch += figmaLink;

    images.push({ match: fullMatch, index: baseStart });
  }

  // 按位置排序
  images.sort((a, b) => a.index - b.index);

  return images;
}

/**
 * 构造新的图片 + Figma 链接 HTML
 */
function buildImageWithLink(filename: string, figmaLink: string, layerName: string): string {
  return `<ac:image><ri:attachment ri:filename="${escapeXmlAttr(filename)}" /></ac:image><p><a href="${escapeXmlAttr(figmaLink)}">[Figma] ${escapeXmlText(layerName)}</a></p>`;
}

/**
 * 解析单个图片位置，分离图片 HTML 和 Figma 链接
 */
function parseImageWithLink(html: string): Array<{ imageHtml: string; figmaLinkHtml: string | null; figmaUrl: string | null; index: number }> {
  const results: Array<{ imageHtml: string; figmaLinkHtml: string | null; figmaUrl: string | null; index: number }> = [];

  // 匹配 ac:image 宏
  const acImageRegex = /<ac:image[^>]*>[\s\S]*?<\/ac:image>/gi;
  let match: RegExpExecArray | null;
  while ((match = acImageRegex.exec(html)) !== null) {
    const imageHtml = match[0];
    const endIndex = match.index + imageHtml.length;
    const figmaLinkHtml = findFollowingFigmaLink(html, endIndex);
    const figmaUrl = figmaLinkHtml ? extractFigmaUrl(figmaLinkHtml) : null;
    
    results.push({ imageHtml, figmaLinkHtml, figmaUrl, index: match.index });
  }

  // 匹配 img 标签
  const imgRegex = /<img[^>]*\/?>/gi;
  while ((match = imgRegex.exec(html)) !== null) {
    const imageHtml = match[0];
    const endIndex = match.index + imageHtml.length;
    const figmaLinkHtml = findFollowingFigmaLink(html, endIndex);
    const figmaUrl = figmaLinkHtml ? extractFigmaUrl(figmaLinkHtml) : null;
    
    results.push({ imageHtml, figmaLinkHtml, figmaUrl, index: match.index });
  }

  // 按位置排序
  results.sort((a, b) => a.index - b.index);

  return results;
}

// ─── Main Export ───

/**
 * 获取 PRD 中现有的图片信息（用于预览）
 */
export async function getExistingImages(
  url: string,
  pat: string,
): Promise<PreviewImagesResult> {
  const { baseUrl, pageId } = parseConfluenceUrl(url);

  log.info({ pageId }, 'Getting existing images for preview');

  try {
    const pageInfo = await getPageInfo(baseUrl, pageId, pat);
    const parsed = parseImageWithLink(pageInfo.body);

    const images: ExistingImageInfo[] = parsed.map((item, idx) => ({
      // Try best-effort preview URL extraction from storage HTML.
      // 1) inline src/ri:value, 2) attachment filename fallback.
      index: idx + 1,
      imageHtml: item.imageHtml,
      imagePreviewUrl: (() => {
        const directSrc = extractImageSrc(item.imageHtml);
        if (directSrc) return directSrc;
        const filename = extractAttachmentFilename(item.imageHtml);
        return filename ? buildAttachmentPreviewUrl(baseUrl, pageId, filename) : null;
      })(),
      figmaLink: item.figmaUrl,
    }));

    log.info({ pageId, imageCount: images.length }, 'Found existing images');

    return {
      success: true,
      images,
      pageTitle: pageInfo.title,
    };
  } catch (err) {
    log.warn({ pageId, error: String(err) }, 'Failed to get existing images');
    return {
      success: false,
      images: [],
      pageTitle: '',
      error: err instanceof Error ? err.message : String(err),
    };
  }
}

/**
 * 更新 PRD 中的图片（先复制原页面，在副本上修改，保留原页面不变）
 * @param url - Confluence 页面 URL
 * @param pat - Personal Access Token
 * @param screenshots - 截图数据数组
 */
export async function updatePrdImages(
  url: string,
  pat: string,
  screenshots: ScreenshotData[],
): Promise<UpdatePrdResult> {
  const { baseUrl, pageId } = parseConfluenceUrl(url);

  log.info({ pageId, screenshotCount: screenshots.length }, 'Starting PRD update (copy mode)');

  // 1. 获取原页面信息
  const pageInfo = await getPageInfo(baseUrl, pageId, pat);
  log.info({ pageId, title: pageInfo.title, version: pageInfo.version }, 'Got original page info');

  // 2. 解析现有图片位置
  const existingImages = findImagePositions(pageInfo.body);
  log.info({ pageId, imageCount: existingImages.length }, 'Found existing images');

  // 3. 创建子页面（复制原内容，标题含时间戳避免重名）
  const now = new Date();
  const timeStr = now.toISOString().replace(/[-:T]/g, '').slice(0, 14); // YYYYMMDDHHmmss
  const newTitle = `${pageInfo.title} - Design Update ${timeStr}`;

  const childPage = await createChildPage(
    baseUrl,
    pageId,
    pageInfo.spaceKey,
    pat,
    newTitle,
    pageInfo.body,
  );
  log.info({ parentPageId: pageId, childPageId: childPage.id, newTitle }, 'Child page created');

  // 4. 上传截图为附件（到子页面）
  const uploadedFiles: Array<{ filename: string; figmaLink: string; name: string }> = [];
  const timestamp = Date.now();

  for (let i = 0; i < screenshots.length; i++) {
    const ss = screenshots[i];
    const filename = `soda-design-${i + 1}-${timestamp}.png`;

    try {
      await uploadAttachment(baseUrl, childPage.id, pat, filename, ss.base64);
      uploadedFiles.push({
        filename,
        figmaLink: ss.figmaLink,
        name: ss.name,
      });
    } catch (err) {
      log.warn({ error: String(err), index: i }, 'Failed to upload screenshot, skipping');
    }
  }

  if (uploadedFiles.length === 0) {
    return {
      success: false,
      updatedCount: 0,
      pageUrl: childPage.url,
      newPageId: childPage.id,
      error: '所有截图上传失败',
    };
  }

  // 5. 按顺序替换图片
  let newBody = pageInfo.body;
  const replaceCount = Math.min(uploadedFiles.length, existingImages.length);

  // 从后往前替换，避免索引偏移
  for (let i = replaceCount - 1; i >= 0; i--) {
    const img = existingImages[i];
    const file = uploadedFiles[i];
    const newContent = buildImageWithLink(file.filename, file.figmaLink, file.name);

    newBody = newBody.slice(0, img.index) + newContent + newBody.slice(img.index + img.match.length);
  }

  // 6. 更新子页面内容
  await updatePageContent(baseUrl, childPage.id, pat, newTitle, newBody, 1);

  log.info({ parentPageId: pageId, childPageId: childPage.id, replacedCount: replaceCount }, 'PRD update completed (copy mode)');

  return {
    success: true,
    updatedCount: replaceCount,
    pageUrl: childPage.url,
    newPageId: childPage.id,
  };
}
